src/geometry/ArborX_GeometryTraits.hpp

```diff
@@ -0,0 +1,65 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_GEOMETRY_TRAITS_HPP
+#define ARBORX_GEOMETRY_TRAITS_HPP
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+
+namespace Experimental
+{
+
+struct PointTag
+{};
+
+struct BoxTag
+{};
+
+struct SphereTag
+{};
+
+template <typename Geometry, typename Enable = void>
+struct GeometryTraits
+{
+  using not_specialized = void; // tag to detect existence of a specialization
+};
+
+template <typename Traits>
+using GeometryTraitsTagArchetypeAlias = typename Traits::tag;
```

Comment:
I would not do that because it is "intrusive".  User types would have to add that tag directly into their geometry type.
```C++
struct MyPoint {
  float x;
  float y;
  float z;
  using tag = ArborX::Experimental::PointTag;  // user forced to inject ArborX symbol into his code
};
```
Prefer something like
```C++
struct MyPoint {
  float x;
  float y;
  float z;
};
template <> struct ArborX::Experimental::GeometryTraits<MyPoint> {
  using tag = PointTag;
  // ...
}
```
I haven't made up my mind yet whether I like better individual traits rather than a monolithic one
```C++
template <> struct ArborX::Experimental::GeometryTraits<MyPoint> {
  using tag = PointTag;
  static constexpr int dim = 3;
  // ...
}
```
vs
```C++
template <> struct ArborX::GeometryTag<MyPoint> {
  using type = PointTag;
};
template <> struct ArborX::GeometryDimension<MyPoint> {
  static constexpr int value = 3;
};
// ...

Reply:
> I would not do that because it is "intrusive". User types would have to add that tag directly into their geometry type.

Not into the geometry type, but in their specialization of `GeometryTraits`, right? Like your second version. That's exactly what I did with `ArborX::Point`, the `tag` is inside the specialization.

Reply:
> I haven't made up my mind yet whether I like better individual traits rather than a monolithic one

Seems like boost is using [individual ones](https://www.boost.org/doc/libs/1_76_0/libs/geometry/doc/html/index.html).